pd/src/state/reader.rs

```diff
@@ -476,6 +477,108 @@ impl Reader {
+    ///
+    /// If `validators` is `Some`, only notes which were associated with an undelegation from some
+    /// validator in that set will be returned. (This is more efficient than filtering after
+    /// receiving he stream, because the database is performing the filtration.)
+    pub fn quarantined_notes<'a>(
+        &self,
+        maximum_unbonding_height: Option<u64>,
+        validators: Option<impl IntoIterator<Item = impl Borrow<&'a IdentityKey>>>,
+    ) -> impl Stream<Item = Result<(IdentityKey, note::Commitment, NoteData)>> + Send + Unpin + '_
+    {
+        // Should we list outputs from all validators?
+        let all_validators = validators.is_none();
+
+        // If not, what's the list of validator identities (as bytes) to filter for?
+        let validator_list = validators
+            .map(|v| v.into_iter().map(|i| i.borrow().encode_to_vec()))
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>();
+
+        query!(
+            "SELECT validator_identity_key, note_commitment, ephemeral_key, encrypted_note, transaction_id
+            FROM quarantined_notes
+            WHERE
+                unbonding_height <= $1 AND
+                ($2 OR validator_identity_key = ANY($3))",
+            maximum_unbonding_height.unwrap_or(u64::MAX) as i64,
+            all_validators,
+            &validator_list,
+        )
+        .fetch(&self.pool)
+        .map_err(Into::into)
+        .map(|result| {
+            result
+                .and_then(|row| {
+                    Ok::<_, anyhow::Error>((
+                        IdentityKey::decode(&*row.validator_identity_key)?,
+                        note::Commitment::try_from(&*row.note_commitment)?,
+                        NoteData {
+                            ephemeral_key: row.ephemeral_key[..].try_into()?,
+                            encrypted_note: row.encrypted_note[..].try_into()?,
+                            transaction_id: row.transaction_id[..].try_into()?,
+                        },
+                    ))
+                })
+                .map_err(Into::into)
+        })
+    }
+
+    /// Retrieve a stream of quarantined nullifiers, paired with the validator identity key with
+    /// which they are associated.
+    ///
+    /// If `maximum_unbonding_height` is `Some`, only nullifiers whose unbonding height is less than or
+    /// equal to that height will be returned.
+    ///
+    /// If `validators` is `Some`, only nullifiers which were associated with an undelegation from some
+    /// validator in that set will be returned. (This is more efficient than filtering after
+    /// receiving he stream, because the database is performing the filtration.)
+    pub fn quarantined_nullifiers<'a>(
+        &self,
+        maximum_unbonding_height: Option<u64>,
+        validators: Option<impl IntoIterator<Item = impl Borrow<&'a IdentityKey>>>,
+    ) -> impl Stream<Item = Result<(IdentityKey, Nullifier)>> + Send + Unpin + '_ {
```

Comment:
This function is called during slashing and unbonding of notes; it's generalized so that it is applicable in both situations.